rasa/utils/tensorflow/model_data_utils.py

```diff
@@ -24,12 +24,123 @@
 TAG_ID_ORIGIN = "tag_id_origin"
 
 
+def extract_attribute_features_from_message(
+    message: Message, attribute: Text, featurizers: Optional[List[Text]] = None,
+) -> Dict[Tuple[Text, Text], Union[scipy.sparse.spmatrix, np.ndarray]]:
+    """Extracts and combines features from the given messages.
+
+    Args:
+        message: any message
+        attribute: attribute for which features should be collected
+        featurizers: the list of featurizers to consider
+    Returns:
+        a dictionary mapping keys that take the form
+        `([True|False],[SEQUENCE|SENTENCE])' to a sparse or dense matrix
+        where True or False indicates that the matrix is sparse or dense, respectively
+    """
+    (sparse_sequence_features, sparse_sentence_features,) = message.get_sparse_features(
+        attribute, featurizers
+    )
+    dense_sequence_features, dense_sentence_features = message.get_dense_features(
+        attribute, featurizers
+    )
+
+    if dense_sequence_features is not None and sparse_sequence_features is not None:
+        if (
+            dense_sequence_features.features.shape[0]
+            != sparse_sequence_features.features.shape[0]
+        ):

```

Comment:
Optional: Looks like this can just be combined in one `if` construct?
(Optional because everyone has their own way of writing these things for better readability 😅 )

Reply:
I didn't write this and I agree ^^

Reply:
More high-level: This is also a kind of assertion that seems just as misplaced as the assertions on sequence/sentence level feature presence and example per label presence that I added here, isn't it? I think long term it might be good to extend the Featurizer interface so that we can more easily check whether what they produce (i.e. which types are there and what's the dimension/type of features produced). What do you think?